src/masternodes/order.cpp

```diff
@@ -0,0 +1,120 @@
+const unsigned char COrderView::OrderCloseTx  ::prefix = 'C';
+
+std::unique_ptr<COrderView::COrderImpl> COrderView::GetOrderByCreationTx(const uint256 & txid) const
+{
+    auto tokenPair = ReadBy<OrderCreationTxId, TokenPair>(txid);
+    if (tokenPair)
+    {
+        auto orderImpl = ReadBy<OrderCreationTx,COrderImpl>(std::make_pair(*tokenPair, txid));
+        if (orderImpl)
+            return MakeUnique<COrderImpl>(*orderImpl);
+    }
+    return (nullptr);
+}
+
+ResVal<uint256> COrderView::CreateOrder(const COrderImpl& order)
+{
+    //this should not happen, but for sure
+    if (GetOrderByCreationTx(order.creationTx)) {
+        return Res::Err("order with creation tx %s already exists!", order.creationTx.GetHex());
+    }
+    TokenPair pair(order.idTokenFrom, order.idTokenTo);
+    TokenPairKey key(pair, order.creationTx);
+    WriteBy<OrderCreationTx>(key, order);
+    WriteBy<OrderCreationTxId>(order.creationTx, pair);
+
+    return {order.creationTx, Res::Ok()};
+}
+
+ResVal<uint256> COrderView::CloseOrderTx(const COrderImpl& order)
+{
+    TokenPairKey key(std::make_pair(order.idTokenFrom, order.idTokenTo), order.creationTx);
+    EraseBy<OrderCreationTx>(key);
+    WriteBy<OrderCloseTx>(key, order);
+    return {order.creationTx, Res::Ok()};
+}
+
+void COrderView::ForEachOrder(std::function<bool (TokenPairKey const &, CLazySerialize<COrderImpl>)> callback, TokenPair const & pair)
+{
+    TokenPairKey start(pair, uint256());
+    ForEach<OrderCreationTx,TokenPairKey,COrderImpl>([&start,&callback] (TokenPairKey const &key, CLazySerialize<COrderImpl> orderImpl) {
+        return callback(key, orderImpl);
+    }, start);
+}
+
+std::unique_ptr<COrderView::CFulfillOrderImpl> COrderView::GetFulfillOrderByCreationTx(const uint256 & txid) const
+{
+    auto ordertxid = ReadBy<FulfillOrderTxid, uint256>(txid);
+    if (ordertxid && !ordertxid->IsNull())
+    {
+        auto fillorderImpl = ReadBy<FulfillCreationTx,CFulfillOrderImpl>(std::make_pair(*ordertxid, txid));
+        if (fillorderImpl)
+            return MakeUnique<CFulfillOrderImpl>(*fillorderImpl);
+        return (nullptr);
+    }
+    return (nullptr);
+}
+
+ResVal<uint256> COrderView::FulfillOrder(const CFulfillOrderImpl & fillorder, const COrderImpl & order)
+{
+    //this should not happen, but for sure
+    if (GetFulfillOrderByCreationTx(fillorder.creationTx)) {
+        return Res::Err("fillorder with creation tx %s already exists!", fillorder.creationTx.GetHex());
+    }
+
+    WriteBy<FulfillOrderTxid>(fillorder.creationTx, fillorder.orderTx);
+    WriteBy<FulfillCreationTx>(std::make_pair(fillorder.orderTx,fillorder.creationTx), fillorder);
+
+    if (order.closeHeight > -1) this->CloseOrderTx(order);
+    else
+    {
+        TokenPair pair(order.idTokenFrom, order.idTokenTo);
+        TokenPairKey key(pair, order.creationTx);
+        WriteBy<OrderCreationTx>(key, order);
+    }
+
+    return {fillorder.creationTx, Res::Ok()};
+}
```

Comment:
Is it necessary to check the full order price match with the ask price?

Reply:
Totally forgot to take into account order price. I've added it now in `amountToFill` field which is already calculated in tokenTo amount necessary.